File: problems/leetcode/0000/1/1.ts
function twoSum(nums: number[], target: number): number[] {
    const numMap = getNumberMap(nums);

    return target % 2 === 0 && numMap.get(target / 2)?.length === 2 ?
        nums
            .reduce((acc, curr, index) => curr === target / 2 ? [ ...acc, index ] : [ ...acc ], [] as number[])
        : nums
            .map((num) => (target % 2 === 1 || num !== target / 2) && numMap.has(target - num))
            .reduce((acc, curr, index) => curr === true ? [ ...acc, index ] : [ ...acc ], [] as number[]);
};

const getNumberMap = (nums: number[]): Map<number, number[]> => {
    const numMap = new Map<number, number[]>();
    nums.forEach((num, index) => numMap.get(num) ? numMap.get(num)?.push(index) : numMap.set(num, [index]));

    return numMap;
}

File: problems/leetcode/0000/1/1.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut countMap: HashMap<i32, i32> = HashMap::new();
        let mut answer: Vec<i32> = vec![];

        for (index, num) in nums.iter().enumerate() {
            if countMap.contains_key(&(target - num)) {
                answer.push(*countMap.get(&(target - num)).unwrap());
                answer.push(index as i32);
                return answer;
            } else {
                countMap.insert(*num, index as i32);
            }
        }

        return answer;
    }
}
